RadixWallet/Clients/FactorSourcesClient/FactorSourcesClient+Live.swift

```diff
@@ -28,44 +28,107 @@ extension FactorSourcesClient: DependencyKey {
+					loggerGlobal.critical("Failed to save factor source, error: \(error)")
+					if let idForMnemonicToDelete = try? factorSourceID.extract(as: FactorSourceID.FromHash.self) {
+						// We were unlucky, failed to update Profile, thus best to undo the saving of
+						// the mnemonic in keychain (if we can).
+						try? secureStorageClient.deleteMnemonicByFactorSourceID(idForMnemonicToDelete)
+					}
+					throw error
+				}
+			}
+
+			return factorSourceID
+		}
+
 		return Self(
 			getCurrentNetworkID: {
 				await profileStore.profile.networkID
 			},
-			getFactorSources: getFactorSources,
-			factorSourcesAsyncSequence: {
-				await profileStore.factorSourcesValues()
-			},
-			addPrivateHDFactorSource: { request in
-				let factorSource = request.factorSource
-
-				switch factorSource {
-				case let .device(deviceFactorSource):
-					try secureStorageClient.saveMnemonicForFactorSource(.init(mnemonicWithPassphrase: request.mnemonicWithPasshprase, factorSource: deviceFactorSource))
-				default:
-					loggerGlobal.notice("Saving of non device private HD factor source not permitted, kind is: \(factorSource.kind)")
-				}
-				let factorSourceID = factorSource.id
-
-				/// We only need to save olympia mnemonics into Profile, the Babylon ones
-				/// already exist in profile, and this function is used only to save the
-				/// imported mnemonic into keychain (done above).
-				if request.saveIntoProfile {
-					do {
-						try await saveFactorSource(factorSource)
-					} catch {
-						loggerGlobal.critical("Failed to save factor source, error: \(error)")
-						if let idForMnemonicToDelete = try? factorSourceID.extract(as: FactorSourceID.FromHash.self) {
-							// We were unlucky, failed to update Profile, thus best to undo the saving of
-							// the mnemonic in keychain (if we can).
-							try? secureStorageClient.deleteMnemonicByFactorSourceID(idForMnemonicToDelete)
-						}
-						throw error
+			getMainDeviceFactorSource: {
+				let sources = try await getFactorSources()
+					.filter { $0.factorSourceKind == .device && !$0.supportsOlympia }
+					.map { try $0.extract(as: DeviceFactorSource.self) }
+
+				if let explicitMain = sources.first(where: { $0.isExplicitMain }) {
+					return explicitMain
+				} else {
+					if sources.count == 0 {
+						let errorMessage = "BAD IMPL found no babylon device factor source"
+						loggerGlobal.critical(.init(stringLiteral: errorMessage))
+						assertionFailure(errorMessage)
+						throw FactorSourceNotFound()
+					} else if sources.count > 1 {
+						let errorMessage = "BAD IMPL found more than 1 implicit main babylon device factor sources"
+						loggerGlobal.critical(.init(stringLiteral: errorMessage))
+						assertionFailure(errorMessage)
+						let dateSorted = sources.sorted(by: { $0.addedOn < $1.addedOn })
+						return dateSorted.first! // best we can do
+					} else {
+						return sources[0] // found implicit one
 					}
 				}
+			},
+			createNewMainDeviceFactorSource: {
```

Comment:
unfortunately this cannot be used by ProfileStore, since this is async, and ProfileStore needs a sync.